contrib/babelfishpg_common/src/varbinary.c

```diff
@@ -693,6 +699,82 @@ varcharvarbinary(PG_FUNCTION_ARGS)
 		FlushErrorState();
 		MemoryContextSwitchTo(ectx);
 
+		ereport(ERROR,
+			(errcode(ERRCODE_INTERNAL_ERROR),
+				errmsg("Failed to convert from data type varchar to varbinary, %s",
+				errorData->message)));
+	}
+	PG_END_TRY();
+
+	/* 
+	 * If typmod is -1 (or invalid), use the actual length
+	 * Length should be checked after encoding into server encoding
+	 */
+	if (typmod < (int32) VARHDRSZ)
+		maxlen = encodedByteLen;
+	else
+		maxlen = typmod - VARHDRSZ;
+
+	if (encodedByteLen > maxlen)
+		encodedByteLen = maxlen;
+	result = (bytea *) palloc(encodedByteLen + VARHDRSZ);
+	SET_VARSIZE(result, encodedByteLen + VARHDRSZ);
+
+	rp = VARDATA(result);
+	memcpy(rp, encoded_data, encodedByteLen);
+
+	PG_RETURN_BYTEA_P(result);
+}
+
+/* 
+ * For nvarchar we need to convert the input string to UTF-16 encoding irrespective of input encoding
+ * So the source string is in UTF-8 encoding, we will convert it to UTF-16 encoding
+ */
+Datum
+nvarcharvarbinary(PG_FUNCTION_ARGS)
+{
+	VarChar         *source = PG_GETARG_VARCHAR_PP(0);
+	char	        *data = VARDATA_ANY(source);		/* Source string is UTF-8 */
+	char	        *encoded_data;
+	char	        *rp;
+	size_t		len = VARSIZE_ANY_EXHDR(source);
+	int32		typmod = PG_GETARG_INT32(1);
+	bool		isExplicit = PG_GETARG_BOOL(2);
+	int32		maxlen;
+	bytea	        *result;
+	int		encodedByteLen;
+	StringInfoData  s; 
+	MemoryContext   ccxt = CurrentMemoryContext;
+
+	if (!isExplicit)
+		ereport(ERROR,
```

Comment:
Do we have test cases for this scenario?

Reply:
https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3072/files#diff-fff75a029d373c427177cfcc5166d287721a02b78cbf1aeae54b868ad6d7bc9eR216